instant-xml-macros/src/lib.rs

```diff
@@ -59,12 +61,66 @@ impl FieldMeta {
                 MetaItem::Attribute => meta.attribute = true,
                 MetaItem::Ns(ns) => meta.ns = ns,
                 MetaItem::Rename(lit) => meta.rename = Some(lit),
+                MetaItem::Scalar => panic!("attribute 'scalar' invalid in field xml attribute"),
             }
         }
         meta
     }
 }
 
+#[derive(Debug, Default)]
+struct VariantMeta {
+    serialize_as: TokenStream,
+}
+
+impl VariantMeta {
+    fn from_variant(input: &syn::Variant) -> VariantMeta {
+        if !input.fields.is_empty() {
+            panic!("only unit enum variants are permitted!")
+        }
+
+        let mut rename = None;
+        for item in meta_items(&input.attrs) {
+            match item {
+                MetaItem::Attribute => panic!("attribute 'attribute' is invalid for enum variants"),
+                MetaItem::Ns(_ns) => panic!("attribute 'ns' is invalid for enum variants"),
+                MetaItem::Rename(lit) => rename = Some(lit.to_token_stream()),
+                MetaItem::Scalar => panic!("attribute 'scalar' is invalid for enum variants"),
+            }
+        }
+
+        let discriminant = match input.discriminant {
+            Some((
+                _,
+                syn::Expr::Lit(syn::ExprLit {
+                    lit: syn::Lit::Str(ref lit),
+                    ..
+                }),
+            )) => Some(lit.to_token_stream()),
+            Some((
+                _,
+                syn::Expr::Lit(syn::ExprLit {
+                    lit: syn::Lit::Int(ref lit),
+                    ..
+                }),
+            )) => Some(lit.base10_digits().to_token_stream()),
+            Some((_, _)) => panic!("invalid field discriminant value!"),
+            None => None,
+        };
+
+        if discriminant.is_some() && rename.is_some() {
+            panic!("conflicting `rename` attribute and variant discriminant!")
+        }
+
+        let serialize_as = rename
+            .or(discriminant)
+            .map(|lit| lit.into_token_stream())
+            .unwrap_or_else(|| input.ident.to_string().to_token_stream());
+
+        VariantMeta { serialize_as }
```

Comment:
I'd prefer to inline the `serialize_as` value here as a `match rename.or(discriminant) { }` (instead of using combinators on both sides).